snuba/datasets/entity.py

```diff
@@ -1,14 +1,27 @@
 from abc import ABC, abstractmethod
-from typing import Mapping, Optional, Sequence
+from typing import Mapping, NamedTuple, Optional, Sequence
 
 from snuba.clickhouse.columns import ColumnSet
 from snuba.pipeline.query_pipeline import QueryPipelineBuilder
+from snuba.datasets.entities import EntityKey
 from snuba.datasets.storage import Storage, WritableStorage, WritableTableStorage
+from snuba.query.data_source.join import JoinCondition, JoinModifier, JoinType
 from snuba.query.extensions import QueryExtension
 from snuba.query.processors import QueryProcessor
 from snuba.query.validation import FunctionCallValidator
 
 
+class JoinRelationship(NamedTuple):
+    """
+    Represents the one way relationship between the owning Entity and another entity.
+    """
+
+    rhs_entity: EntityKey
+    keys: Sequence[JoinCondition]
+    join_type: JoinType
+    join_modifier: Optional[JoinModifier] = None
```

Comment:
I don't think this is a structural property of the relationship but a property of the query we run. We can run a standard join or a semi join between spans and transactions as a standard join or as a semi join depending on the query.

I think a better way to classify a relationship structurally in the data model in a way that is helpful when optimizing the query is to define whether the relationship is 1:1, 1:n, n:1, [0..1]: n etc. Those are structural properties. 

The parser can simply decide whether to build INNER vs LEFT depending on the cardinality but it does not need to deal with SEMI join, which is a join optimization matter.